### PR TITLE
ZoneCog: tiered hot/cold hypergraph storage (Phase E.2, issue #81)

### PR DESCRIPTION
**Pull request by drzo**
## Description

Continues the ZoneCog implementation plan tracked in #76. With Phase A (#77) and Phase B (#78) both closed, this picks up the next item in the epic's recommended order — Phase E, Enhanced Persistence Layer (#81) — specifically the "Hybrid Storage Strategy" (E.2) and part of the export enhancements (E.3):

- `HypergraphPersistenceService` gains tiered hot/cold storage: `archiveLowSalienceNodes(threshold?)` moves nodes whose `salience_score` is below a threshold (default `0.05`) out of the live in-memory hypergraph into a new IndexedDB cold tier (`archiveNodes`/`archiveLinks` object stores, DB bumped to version 2). A link is archived alongside its nodes only when *every* one of its outgoing ids is also being archived in the same pass, so links with a remaining hot endpoint stay live.
- `restoreArchivedNode(nodeId)` lazily restores a single archived node — and the links it directly references — back into the live hypergraph on demand, without materializing the rest of the cold tier.
- `listArchivedNodes()` browses cold-tier contents without loading them into memory.
- `getStats()` now reports `archivedNodeCount`.
- Two new Command Palette actions: **Zone-Cog: Archive Low-Salience Nodes** and **Zone-Cog: Restore Archived Node** (quick-pick over the archived set).

While auditing E.3 ("Export/Import Enhancements") I found Neo4j Cypher export and OpenCog AtomSpace Scheme export were already implemented in a prior PR but not reflected as complete in `docs/ZONECOG_ROADMAP.md`; that doc is updated accordingly, and the remaining E.1 (RocksDB-via-WASM backend) and the rest of E.3 (incremental backup/restore, cloud storage) are called out explicitly as future work — RocksDB in particular would need a new WASM dependency and build-system integration that's out of scope here, consistent with the epic's own risk assessment which already treats IndexedDB as the fallback.

Closes part of #81.

## Code Changes Checklist

- [x] New or updated **unit tests** added — 7 new tests covering archival (including partial-link-ownership), lazy restore, and stats reporting, plus IndexedDB mock support for `get`/`delete` and the two new object stores
- [ ] All existing tests pass (`npm run test`) — this sandbox has no `node_modules` available to run the full suite; verified instead with a scoped `tsc --noEmit` pass restricted to the touched files (zero errors) and careful manual review against the existing test harness patterns. Please confirm green in CI.
- [x] Code follows [contributing guidelines](https://github.com/microsoft/azuredatastudio/blob/main/CONTRIBUTING.md)
- [x] No UI regressions introduced — additive service methods and two new opt-in Command Palette actions only
- [ ] Logging/telemetry updated if applicable — n/a, uses existing `ILogService`/membrane activity recording patterns
- [x] Cross-platform support checked (Windows, macOS, Linux if relevant) — browser/IndexedDB-only code path, no platform-specific behavior

## Reviewers: [Please read our reviewer guidelines](https://github.com/microsoft/azuredatastudio/blob/main/.github/REVIEW_GUIDELINES.md)

---
_Generated by [Claude Code](https://claude.ai/code/session_018vEwvw9WpfuFHZAcKXAump)_

---

> [!NOTE]
> **Medium Risk**
> Changes hypergraph persistence semantics (IndexedDB schema v2, in-memory removals, restore/link edge cases); mitigated by serialization and extensive tests, but data-loss or inconsistency bugs would affect stored knowledge graphs.
> 
> **Overview**
> Adds **Phase E.2 hybrid storage** to `HypergraphPersistenceService`: low-salience nodes move from the live graph into IndexedDB **cold** stores (`archiveNodes` / `archiveLinks`, DB version 2), with `archiveLowSalienceNodes()`, lazy `restoreArchivedNode()`, and `listArchivedNodes()`. Links archive only when every outgoing endpoint is archived in the same pass; restore keeps shared cold links until no archived node still references them, and avoids writing links into the hot tier while an endpoint remains cold.
> 
> **Save**, **load**, **clear**, and archive/restore run through a **`_serialize` write queue** so concurrent auto-save cannot resurrect archived nodes. `getStats()` adds `archivedNodeCount`; `clearStorage()` wipes cold tiers too.
> 
> Two Command Palette actions expose archive and restore (quick-pick). `docs/ZONECOG_ROADMAP.md` documents completed Cypher/Scheme export and tiered storage; RocksDB and incremental/cloud backup stay future work. Broad unit tests cover tiering, hot-store durability, and save/archive races.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f7ef3e4ce9a72db93284a40116b793e8fc0d1280. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts [hunk 1]
```diff
@@ -480,6 +517,91 @@ export class HypergraphPersistenceService extends Disposable implements IHypergr
...
+		const archiveLinkStore = tx.objectStore(STORE_ARCHIVE_LINKS);
+		for (const node of toArchive) { await idbPut(archiveNodeStore, node); }
+		for (const link of linksToArchive) { await idbPut(archiveLinkStore, link); }
+		await new Promise<void>((resolve, reject) => {
+			tx.oncomplete = () => resolve();
+			tx.onerror = () => reject(tx.error);
+		});
+
+		for (const link of linksToArchive) { this.hypergraphStore.removeLink(link.id); }
+		for (const node of toArchive) { this.hypergraphStore.removeNode(node.id); }
```
[the remark this answers is not in the document]
Comment (drzo) [reply src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts]: Fixed in fd97b15a: `archiveLowSalienceNodes` now deletes the archived node/link records from the hot `nodes`/`links` IndexedDB stores in the same transaction that writes them into the cold tier, so `load()` can no longer resurrect them.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts [hunk 2]
```diff
@@ -480,6 +517,91 @@ export class HypergraphPersistenceService extends Disposable implements IHypergr
...
+			}
+		}
+
+		const tx = db.transaction([STORE_ARCHIVE_NODES, STORE_ARCHIVE_LINKS], 'readwrite');
+		await idbDelete(tx.objectStore(STORE_ARCHIVE_NODES), nodeId);
+		for (const link of links) { await idbDelete(tx.objectStore(STORE_ARCHIVE_LINKS), link.id); }
+		await new Promise<void>((resolve, reject) => {
+			tx.oncomplete = () => resolve();
+			tx.onerror = () => reject(tx.error);
+		});
```
[the remark this answers is not in the document]
Comment (drzo) [reply src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts]: Fixed in fd97b15a: `restoreArchivedNode` now writes the restored node and its links into the hot `nodes`/`links` stores as part of the same transaction that deletes them from cold storage, so the restore is durable without requiring a separate `save()` call before the next `load()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts [hunk 3]
```diff
@@ -480,6 +517,91 @@ export class HypergraphPersistenceService extends Disposable implements IHypergr
...
+		this.hypergraphStore.addNode(node);
+		for (const link of links) {
+			if (!this.hypergraphStore.getLink(link.id)) {
+				this.hypergraphStore.addLink(link);
+			}
+		}
+
+		const tx = db.transaction([STORE_ARCHIVE_NODES, STORE_ARCHIVE_LINKS], 'readwrite');
+		await idbDelete(tx.objectStore(STORE_ARCHIVE_NODES), nodeId);
+		for (const link of links) { await idbDelete(tx.objectStore(STORE_ARCHIVE_LINKS), link.id); }
```
[the remark this answers is not in the document]
Comment (drzo) [reply src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts]: Fixed in fd97b15a: before dropping a link from cold storage, `restoreArchivedNode` now checks whether any *other* still-archived node's `.links` still references it and keeps it archived if so. Added a regression test (`should keep a cold link archived while another archived node still references it`) that restores two nodes sharing one cold link one at a time and asserts the link only leaves archive storage once the last archived owner is restored.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts [hunk 4]
```diff
@@ -440,14 +474,16 @@ export class HypergraphPersistenceService extends Disposable implements IHypergr
 				lastSaveTime: this._lastSaveTime,
 				lastLoadTime: this._lastLoadTime,
 				estimatedBytes: 0,
+				archivedNodeCount: 0,
 			};
 		}
 
 		const db = await this._getDb();
-		const [nodeCount, linkCount, snapshotCount] = await Promise.all([
```
[the remark this answers is not in the document]
Comment (drzo) [reply src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts:447]: Fixed in fd97b15a: `clearStorage()` now includes `archiveNodes`/`archiveLinks` in the transaction and clears them alongside `nodes`/`links`/`snapshots`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts [hunk 5]
```diff
@@ -480,6 +519,122 @@ export class HypergraphPersistenceService extends Disposable implements IHypergr
...
+
+		// Write the restored records into the hot tier and remove the node
+		// (plus any link no other archived node still needs) from the cold
+		// tier, so the restore is durable without requiring a separate
+		// save() call before the next load().
+		const tx = db.transaction([STORE_NODES, STORE_LINKS, STORE_ARCHIVE_NODES, STORE_ARCHIVE_LINKS], 'readwrite');
+		await idbPut(tx.objectStore(STORE_NODES), node);
+		for (const link of links) { await idbPut(tx.objectStore(STORE_LINKS), link); }
+		await idbDelete(tx.objectStore(STORE_ARCHIVE_NODES), nodeId);
+		for (const link of linksToDropFromArchive) { await idbDelete(tx.objectStore(STORE_ARCHIVE_LINKS), link.id); }
```
[the remark this answers is not in the document]
Comment (drzo) [reply src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts]: Fixed in 94dbd792: `restoreArchivedNode` now only writes a link into the hot `links` store once every one of its outgoing nodes is actually live (`hotLinks` filter). A link with a still-cold endpoint stays purely in-memory for the session instead, so re-archiving the restored (still low-salience) node can't leave it orphaned in hot storage. Added a regression test that restores one endpoint of a shared link, re-archives it, and asserts the hot `links` store stays empty and `load()` doesn't resurrect the link.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts [hunk 6]
```diff
@@ -480,6 +519,122 @@ export class HypergraphPersistenceService extends Disposable implements IHypergr
...
+		const archiveNodeStore = tx.objectStore(STORE_ARCHIVE_NODES);
+		const archiveLinkStore = tx.objectStore(STORE_ARCHIVE_LINKS);
+		for (const node of toArchive) {
+			await idbPut(archiveNodeStore, node);
+			await idbDelete(nodeStore, node.id);
+		}
+		for (const link of linksToArchive) {
+			await idbPut(archiveLinkStore, link);
+			await idbDelete(linkStore, link.id);
+		}
```
[the remark this answers is not in the document]
Comment (drzo) [reply src/sql/workbench/services/zonecog/browser/hypergraphPersistenceService.ts]: Fixed in 94dbd792: added a `_serialize()` FIFO queue and routed `save()`, `load()`, `clearStorage()`, `archiveLowSalienceNodes()`, and `restoreArchivedNode()` through it, so none of these can interleave. A save queued behind an in-flight archive won't take its in-memory snapshot until the archive has fully committed and updated the live graph, so it can no longer write a stale (pre-archive) snapshot back over the hot tier. Added a regression test that fires `save()` and `archiveLowSalienceNodes()` concurrently (without awaiting the first) and asserts the archived node stays out of the hot store.

---
_Generated by [Claude Code](https://claude.ai/code)_
